File: Luogu/P1915/Sub6.pas
program nemo;
{for case 9}
const max=10000000;
      zero=1e-6;
type
  ty1=record
    w,x,y,p,q:extended;
  end;
  ty2=record
    t,x,y:extended;
    s:longint;
  end;
var
  w,x,y,v,tot_t,now_t,ans,d:extended;
  n,i,j,t,tot:longint;
  f:array[0..10000] of ty1;
  get:array[0..10000] of boolean;
  step:array[0..10000] of ty2;
//============================
function dis(i:longint; x,y:extended):extended;
begin
  exit(sqrt(sqr(x-f[i].x)+sqr(y-f[i].y)));
end;
//============================
begin
  assign(input,'nemo9.in'); reset(input);
  assign(output,'nemo9.out'); rewrite(output);
  read(w,v,tot_t,x,y);
  read(n); ans:=0; tot:=0; now_t:=0;
  fillchar(get,sizeof(get),false);
  for i:=1 to n do read(f[i].w,f[i].x,f[i].y,f[i].p,f[i].q);
  for i:=1 to n do
  begin
    d:=max; t:=0;
    for j:=1 to n do
    if not get[j] and (f[j].w-w<-zero) and (dis(j,x,y)<d)  then
    begin
      t:=j;
      d:=dis(j,x,y);
    end;
    if t=0 then break;
    now_t:=now_t+dis(t,x,y)/v;
    if now_t-tot_t>zero then break;
    get[t]:=true;
    w:=w+f[t].w;
    ans:=ans+f[t].w;
    inc(tot);
    step[tot].t:=now_t;
    step[tot].s:=t;
    step[tot].x:=f[t].x;
    step[tot].y:=f[t].y;
    x:=f[t].x; y:=f[t].y;
  end;
  writeln(tot);
  writeln(ans:0:6);
  for i:=1 to tot do
  writeln(step[i].t:0:6,' ',step[i].x:0:6,' ',step[i].y:0:6,' ',step[i].s);
  close(input); close(output);
end.

File: Luogu/P1915/Sub6.cpp
#include<iostream>
#include<cstdio>
#include<cmath>
#include<cstring>
using namespace std;
#define w w0
#define v V
#define tot_t T
// #define x x0
// #define y y0
#define max Max
const int max=10000000;
const double zero=1e-6;
struct ty1
{
    double w,x,y,p,q;
};
struct ty2
{
    double t,x,y;
    int s;
};
double w,x,y,v,tot_t,now_t,ans,d;
int n,i,j,t,tot;
ty1 f[10001];
bool get[10001];
ty2 step[10001];
double dis(int i,double x,double y)
{
    return (sqrt(sqrt(x-f[i].x)+sqrt(y-f[i].y)));
}
int main()
{
    freopen("nemo9.in","r",stdin);
    freopen("temp.txt","w",stdout);
    scanf("%lf%lf%lf%lf%lf",&w,&v,&tot_t,&x,&y);
    scanf("%d",&n);
    ans=0,tot=0,now_t=0;
    memset(get,false,sizeof(get));
    for(int i=1;i<=n;i++)
        scanf("%lf%lf%lf%lf%lf",&f[i].w,&f[i].x,&f[i].y,&f[i].p,&f[i].q);
    for(int i=1;i<=n;i++)
    {
        d=max;t=0;
        for(int j=1;j<=n;j++)
            if(!get[j]&&f[j].w-w<zero&&dis(j,x,y)<d)
            {
                t=j,d=dis(j,x,y);
            }
        if(t==0) break;
        now_t=now_t+dis(t,x,y)/v;
        if(now_t-tot_t>zero) break;
        get[t]=true;
        w=w+f[t].w;
        ans=ans+f[t].w;
        tot++;
        step[tot].t=now_t;
        step[tot].s=t;
        step[tot].x=f[t].x;
        step[tot].y=f[t].y;
        x=f[t].x,y=f[t].y;
    }
    printf("%d\n",tot);
    printf("%.6lf\n",ans);
    for(int i=1;i<=tot;i++)
        printf("%.6lf %.6lf %.6lf %d\n",step[i].t,step[i].x,step[i].y,step[i].s);
    return 0;
}